File: java/com/google/copybara/hg/HgRepository.java
/*
 * Copyright (C) 2018 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.hg;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.copybara.util.CommandRunner.DEFAULT_TIMEOUT;
import static com.google.copybara.util.RepositoryUtil.validateNotHttp;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Preconditions;
import com.google.common.base.Strings;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Iterables;
import com.google.common.reflect.TypeToken;
import com.google.copybara.exception.CannotResolveRevisionException;
import com.google.copybara.exception.RepoException;
import com.google.copybara.exception.ValidationException;
import com.google.copybara.util.BadExitStatusWithOutputException;
import com.google.copybara.util.CommandOutput;
import com.google.copybara.util.CommandOutputWithStatus;
import com.google.copybara.util.CommandRunner;
import com.google.gson.Gson;
import com.google.gson.JsonParseException;
import com.google.gson.annotations.SerializedName;
import com.google.copybara.shell.Command;
import com.google.copybara.shell.CommandException;
import com.google.re2j.Pattern;
import java.io.IOException;
import java.lang.reflect.Type;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.time.Instant;
import java.time.ZoneId;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import javax.annotation.Nullable;

/**
 * A class for manipulating Hg (Mercurial) repositories
 */
public class HgRepository {

  /**
   * Label to mark the original revision id (Hg SHA-1) for migrated commits.
   */
  static final String HG_ORIGIN_REV_ID = "HgOrigin-RevId";

  private static final Pattern INVALID_HG_REPOSITORY =
      Pattern.compile("abort: repository .+ not found");

  private static final Pattern UNKNOWN_REVISION =
      Pattern.compile("abort: unknown revision '.+'");

  private static final Pattern INVALID_REF_EXPRESSION =
      Pattern.compile("syntax error in revset '.+'");

  /**
   * The location of the {@code .hg} directory.
   */
  private final Path hgDir;
  private final boolean verbose;
  private final Duration fetchTimeout;

  public HgRepository(Path hgDir, boolean verbose, Duration fetchTimeout) {
    this.hgDir = checkNotNull(hgDir);
    this.verbose = verbose;
    this.fetchTimeout = checkNotNull(fetchTimeout);
  }

  /**
   * Initializes a new hg repository
   * @return the new HgRepository
   * @throws RepoException if the directory cannot be created
   */
  public HgRepository init() throws RepoException {
    try {
      Files.createDirectories(hgDir);
    } catch (IOException e) {
      throw new RepoException("Cannot create directory: " + e.getMessage(), e);
    }
    hg(hgDir, ImmutableList.of("init"), DEFAULT_TIMEOUT);
    return this;
  }

  /**
   * Finds all changes from a repository at {@code url} and adds to the current repository.
   * Defaults to forced pull.
   *
   * <p>This does not update the copy of the project in the working directory.
   *
   * @param url remote hg repository url
   */
  void pullAll(String url) throws RepoException, ValidationException {
    pull(url, /*force*/ true, /*ref*/ null);
  }

  /**
   * Finds a single reference from a repository at {@code url} and adds to the current repository.
   * Defaults to forced pull.
   *
   * <p>This does not update the copy of the project in the working directory.
   *
   * @param url remote hg repository url
   * @param ref the remote revision to add
   */
  void pullFromRef(String url, String ref) throws RepoException, ValidationException {
    pull(url, /*force*/ true, /*ref*/ ref);
  }


  public void pull(String url, boolean force, @Nullable String ref)
      throws RepoException, ValidationException {
    ImmutableList.Builder<String> builder = ImmutableList.builder();

    builder.add("pull", validateNotHttp(url));

    if (force) {
      builder.add("--force");
    }

    if (!Strings.isNullOrEmpty(ref)) {
      builder.add("--rev", ref);
    }

    try {
      hg(hgDir, builder.build(), fetchTimeout);
    } catch (RepoException e) {
      if (INVALID_HG_REPOSITORY.matcher(e.getMessage()).find()){
        throw new ValidationException("Repository not found: " + e.getMessage());
      }
      if (UNKNOWN_REVISION.matcher(e.getMessage()).find()) {
        throw new ValidationException("Unknown revision: " + e.getMessage());
      }
      throw e;
    }
  }

  /**
   * Updates the working directory to the revision given at {@code ref} in the repository
   * and discards local changes.
   */
  CommandOutput cleanUpdate(String ref) throws RepoException {
      return hg(hgDir, "update", ref, "--clean");
  }

  /**
   * Returns a revision object given a reference.
   *
   * <p> A reference can be any of the following:
   * <ul>
   *   <li> A global identifier for a revision. Example: f4e0e692208520203de05557244e573e981f6c72
   *   <li> A local identifier for a revision in the repository. Example: 1
   *   <li> A bookmark in the repository.
   *   <li> A branch in the repository, which returns the tip of that branch. Example: default
   *   <li> A tag in the repository. Example: tip
   * </ul>
   *
   */
  public HgRevision identify(String reference)
      throws RepoException, CannotResolveRevisionException {
    try {
      CommandOutput commandOutput =
          hg(hgDir, "identify", "--template", "{node}\n", "--id", "--rev", reference);

      String globalId = commandOutput.getStdout().trim();
      return new HgRevision(globalId, reference);
    } catch (RepoException e) {
      if (UNKNOWN_REVISION.matcher(e.getMessage()).find()){
        throw new CannotResolveRevisionException(
            String.format("Unknown revision: %s", e.getMessage()));
      }
      throw e;
    }
  }

  /**
   * Creates an unversioned archive of the current working directory and subrepositories
   * in the location {@code archivePath}.
   */
  void archive(String archivePath) throws RepoException {
    hg(hgDir, "archive", archivePath, "--type", "files", "--subrepos");
  }

  /**
   * Creates a log command.
   */
  public LogCmd log() {
    return LogCmd.create(this);
  }

  /**
   * Invokes {@code hg} in the directory given by {@code cwd} against this repository and returns
   * the {@link CommandOutput} if the command execution was successful.
   *
   * <p>Only to be used externally for testing.
   *
   * @param cwd the directory in which to execute the command
   * @param params the argv to pass to Hg, excluding the initial {@code hg}
   */
  @VisibleForTesting
  public CommandOutput hg(Path cwd, String... params) throws RepoException {
    return hg(cwd, Arrays.asList(params), DEFAULT_TIMEOUT);
  }

  private CommandOutput hg(Path cwd, Iterable<String> params, Duration timeout)
      throws RepoException {
    try {
      return executeHg(cwd, params, -1, timeout);
    } catch (BadExitStatusWithOutputException e) {
      throw new RepoException(String.format("Error executing hg: %s", e.getOutput().getStderr()));
    } catch (CommandException e) {
      throw new RepoException(String.format("Error executing hg: %s", e.getMessage()));
    }
  }

  public Path getHgDir() {
    return hgDir;
  }

  private CommandOutputWithStatus executeHg(Path cwd, Iterable<String> params,
      int maxLogLines, Duration timeout) throws CommandException {
    List<String> allParams = new ArrayList<>(Iterables.size(params) + 1);
    allParams.add("hg"); //TODO(jlliu): resolve Hg binary here
    Iterables.addAll(allParams, params);
    Command cmd = new Command(
        Iterables.toArray(allParams, String.class), null, cwd.toFile());
        //TODO(jlliu): have environment vars
    CommandRunner runner = new CommandRunner(cmd, timeout).withVerbose(verbose);
    return
        maxLogLines >= 0 ? runner.withMaxStdOutLogLines(maxLogLines).execute() : runner.execute();
  }

  /**
   * An object that can perform a "hg log" operation on a repository and returns a list of
   * {@link HgLogEntry}.
   */
  public static class LogCmd {

    private final HgRepository repo;

    private final int limit;

    /** Branch to limit the query from. Defaults to all branches if null.*/
    @Nullable
    private final String branch;

    @Nullable
    private final String referenceExpression;

    @Nullable
    private final String keyword;

    private LogCmd(HgRepository repo, int limit, @Nullable String branch,
        @Nullable String referenceExpression, @Nullable String keyword) {
      this.repo = repo;
      this.limit = limit;
      this.branch = branch;
      this.referenceExpression = referenceExpression;
      this.keyword = keyword;
    }

    static LogCmd create(HgRepository repo) {
      return new LogCmd(
          checkNotNull(repo), /*limit*/0, /*branch*/null,
          /*referenceExpression*/ null, /*keyword*/ null);
    }

    /**
     * Limit the query to references that fit the {@code referenceExpression}.
     *
     * <p>The expression must be in a format that is supported by Mercurial. Mercurial supports a
     * number of operators: for example, x::y represents all revisions that are descendants of x and
     * ancestors of y, including x and y.
     */
    LogCmd withReferenceExpression(String referenceExpression) throws RepoException {
      if (Strings.isNullOrEmpty(referenceExpression.trim())){
        throw new RepoException("Cannot log null or empty reference");
      }
      return new LogCmd(repo, limit, branch, referenceExpression.trim(), keyword);
    }


    /**
     * Limit the query to {@code limit} results. Should be > 0.
     */
    public LogCmd withLimit(int limit) {
      Preconditions.checkArgument(limit > 0);
      return new LogCmd(repo, limit, branch, referenceExpression, keyword);
    }

    /**
     * Only query for revisions from the branch {@code branch}.
     */
    LogCmd withBranch(String branch) {
      return new LogCmd(repo, limit, branch, referenceExpression, keyword);
    }

    /**
     * Only query for revisions with the keyword {@code keyword}.
     */
    LogCmd withKeyword(String keyword) {
      return new LogCmd(repo, limit, branch, referenceExpression, keyword);
    }

    /**
     * Run "hg log' and return zero or more {@link HgLogEntry}.
     *
     * <p> The order of the log entries is by default in reverse chronological order, where the
     * first element of the list is the most recent commit. However, if a reference expression is
     * provided, log entries will be ordered with respect to the expression. For example, if the
     * expression is "0:tip", entries will be ordered in chronological order, where the first
     * element of the list is the first commit.
     */
    public ImmutableList<HgLogEntry> run() throws RepoException, ValidationException {
      ImmutableList.Builder<String> builder = ImmutableList.builder();
      builder.add("log", "--verbose"); // verbose flag shows files in output

      if (branch != null) {
        builder.add("--branch", branch);
      }

      /* hg requires limit to be a positive integer */
      if (limit > 0) {
        builder.add("--limit", String.valueOf(limit));
      }

      if (referenceExpression != null) {
        builder.add("--rev", referenceExpression);
      }

      if (keyword != null) {
        builder.add("--keyword", keyword);
      }

      builder.add("-Tjson");
      try {
        CommandOutput output = repo.hg(repo.getHgDir(), builder.build(), DEFAULT_TIMEOUT);
        return parseLog(output.getStdout());
      } catch (RepoException e) {
        if (UNKNOWN_REVISION.matcher(e.getMessage()).find()) {
          throw new ValidationException("Unknown revision: " + e.getMessage());
        }
        if (INVALID_REF_EXPRESSION.matcher(e.getMessage()).find()) {
          throw new RepoException("Syntax error in reference expression: " + e.getMessage());
        }
        throw e;
      }
    }

    private ImmutableList<HgLogEntry> parseLog(String log) throws RepoException {
      if (log.isEmpty()) {
        return ImmutableList.of();
      }

      Gson gson = new Gson();
      Type logListType = new TypeToken<List<HgLogEntry>>() {}.getType();

      try {
        List<HgLogEntry> logEntries = gson.fromJson(log.trim(), logListType);
        return ImmutableList.copyOf(logEntries);
      } catch (JsonParseException e) {
        throw new RepoException(String.format("Cannot parse log output: %s", e.getMessage()));
      }
    }
  }

  /**
   * An object that represents a commit as returned by 'hg log'.
   */
  public static class HgLogEntry {
    @SerializedName("node") private final String globalId;
    private final List<String> parents;
    private final String user;
    @SerializedName("date") private final List<String> commitDate;
    private final String branch;
    @SerializedName("desc") private final String description;
    private final List<String> files;

    HgLogEntry(String node, List<String> parents, String user,
        List<String> commitDate, String branch,
        String desc, List<String> files) {
      this.globalId = checkNotNull(node);
      this.parents = ImmutableList.copyOf(parents);
      this.user = user;
      this.commitDate = commitDate;
      this.branch = branch;
      this.description = desc;
      this.files = ImmutableList.copyOf(files);
    }

    public List<String> getParents() {
      return parents;
    }

    public String getUser() {
      return user;
    }

    String getGlobalId() {
      return globalId;
    }

    ZonedDateTime getZonedDate() {
      Instant date = Instant.ofEpochSecond(
          Long.parseLong(commitDate.get(0)));
      ZoneOffset offset = ZoneOffset.ofTotalSeconds(-1 * Integer.parseInt(commitDate.get(1)));
      ZoneId zone = ZoneId.ofOffset("", offset);
      return ZonedDateTime.ofInstant(date, zone);
    }

    public String getBranch() {
      return branch;
    }

    public String getDescription() {
      return description;
    }

    public List<String> getFiles() {
      return files;
    }
  }
}

File: src/Copybara.Core/Hg/HgRepository.cs
/*
 * Copyright (C) 2018 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Immutable;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;
using Copybara.Common;
using Copybara.Exceptions;
using Copybara.Util;

namespace Copybara.Hg;

/// <summary>A class for manipulating Hg (Mercurial) repositories.</summary>
public class HgRepository
{
    /// <summary>Label to mark the original revision id (Hg SHA-1) for migrated commits.</summary>
    internal const string HgOriginRevId = "HgOrigin-RevId";

    private static readonly Regex InvalidHgRepository =
        new("abort: repository .+ not found", RegexOptions.Compiled);

    private static readonly Regex UnknownRevision =
        new("abort: unknown revision '.+'", RegexOptions.Compiled);

    private static readonly Regex InvalidRefExpression =
        new("syntax error in revset '.+'", RegexOptions.Compiled);

    /// <summary>The location of the <c>.hg</c> directory.</summary>
    private readonly string _hgDir;
    private readonly bool _verbose;
    private readonly TimeSpan _repoTimeout;

    public HgRepository(string hgDir, bool verbose, TimeSpan repoTimeout)
    {
        _hgDir = Preconditions.CheckNotNull(hgDir);
        _verbose = verbose;
        _repoTimeout = repoTimeout;
    }

    /// <summary>Initializes a new hg repository.</summary>
    /// <returns>the new HgRepository</returns>
    /// <exception cref="RepoException">if the directory cannot be created</exception>
    public HgRepository Init()
    {
        try
        {
            Directory.CreateDirectory(_hgDir);
        }
        catch (IOException e)
        {
            throw new RepoException("Cannot create directory: " + e.Message, e);
        }

        Hg(_hgDir, new[] { "init" }, CommandRunner.DefaultTimeout);
        return this;
    }

    /// <summary>
    /// Finds all changes from a repository at <paramref name="url"/> and adds to the current
    /// repository. Defaults to forced pull.
    /// </summary>
    internal void PullAll(string url) => Pull(url, force: true, @ref: null);

    /// <summary>
    /// Finds a single reference from a repository at <paramref name="url"/> and adds to the current
    /// repository. Defaults to forced pull.
    /// </summary>
    internal void PullFromRef(string url, string? @ref) => Pull(url, force: true, @ref: @ref);

    public void Pull(string url, bool force, string? @ref)
    {
        var builder = ImmutableArray.CreateBuilder<string>();
        builder.Add("pull");
        builder.Add(ValidateNotHttp(url));

        if (force)
        {
            builder.Add("--force");
        }

        if (!string.IsNullOrEmpty(@ref))
        {
            builder.Add("--rev");
            builder.Add(@ref);
        }

        try
        {
            Hg(_hgDir, builder.ToImmutable(), _repoTimeout);
        }
        catch (RepoException e)
        {
            if (e.Message is not null && InvalidHgRepository.IsMatch(e.Message))
            {
                throw new ValidationException("Repository not found: " + e.Message);
            }

            if (e.Message is not null && UnknownRevision.IsMatch(e.Message))
            {
                throw new ValidationException("Unknown revision: " + e.Message);
            }

            throw;
        }
    }

    /// <summary>
    /// Updates the working directory to the revision given at <paramref name="ref"/> in the
    /// repository and discards local changes.
    /// </summary>
    internal CommandOutput CleanUpdate(string @ref) => Hg(_hgDir, "update", @ref, "--clean");

    /// <summary>Returns a revision object given a reference.</summary>
    public HgRevision Identify(string reference)
    {
        try
        {
            CommandOutput commandOutput =
                Hg(_hgDir, "identify", "--template", "{node}\n", "--id", "--rev", reference);

            string globalId = commandOutput.GetStdout().Trim();
            return new HgRevision(globalId, reference);
        }
        catch (RepoException e)
        {
            if (e.Message is not null && UnknownRevision.IsMatch(e.Message))
            {
                throw new CannotResolveRevisionException($"Unknown revision: {e.Message}");
            }

            throw;
        }
    }

    /// <summary>
    /// Creates an unversioned archive of the current working directory and subrepositories in the
    /// location <paramref name="archivePath"/>.
    /// </summary>
    internal void Archive(string archivePath) =>
        Hg(_hgDir, "archive", archivePath, "--type", "files", "--subrepos");

    /// <summary>Creates a log command.</summary>
    public LogCmd Log() => LogCmd.Create(this);

    /// <summary>
    /// Invokes <c>hg</c> in the directory given by <paramref name="cwd"/> against this repository and
    /// returns the <see cref="CommandOutput"/> if the command execution was successful.
    ///
    /// <para>Only to be used externally for testing.</para>
    /// </summary>
    public CommandOutput Hg(string cwd, params string[] @params) =>
        Hg(cwd, @params, CommandRunner.DefaultTimeout);

    private CommandOutput Hg(string cwd, IEnumerable<string> @params, TimeSpan timeout)
    {
        try
        {
            return ExecuteHg(cwd, @params, -1, timeout);
        }
        catch (BadExitStatusWithOutputException e)
        {
            throw new RepoException($"Error executing hg: {e.GetOutput().GetStderr()}");
        }
        catch (CommandException e)
        {
            throw new RepoException($"Error executing hg: {e.Message}");
        }
    }

    public string GetHgDir() => _hgDir;

    private CommandOutputWithStatus ExecuteHg(
        string cwd, IEnumerable<string> @params, int maxLogLines, TimeSpan timeout)
    {
        var allParams = new List<string> { "hg" }; // TODO(jlliu): resolve Hg binary here
        allParams.AddRange(@params);
        // TODO(jlliu): have environment vars
        var cmd = new Command(allParams.ToArray(), null, cwd);
        CommandRunner runner = new CommandRunner(cmd, timeout).WithVerbose(_verbose);
        return maxLogLines >= 0
            ? runner.WithMaxStdOutLogLines(maxLogLines).Execute()
            : runner.Execute();
    }

    // Port of RepositoryUtil.validateNotHttp.
    private static string ValidateNotHttp(string url)
    {
        ValidationException.CheckCondition(
            !url.StartsWith("http://", StringComparison.Ordinal),
            "URL '{0}' is not valid - should be using https.",
            url);
        return url;
    }

    /// <summary>
    /// An object that can perform a "hg log" operation on a repository and returns a list of
    /// <see cref="HgLogEntry"/>.
    /// </summary>
    public sealed class LogCmd
    {
        private readonly HgRepository _repo;
        private readonly int _limit;

        /// <summary>Branch to limit the query from. Defaults to all branches if null.</summary>
        private readonly string? _branch;

        private readonly string? _referenceExpression;
        private readonly string? _keyword;

        private LogCmd(
            HgRepository repo, int limit, string? branch, string? referenceExpression, string? keyword)
        {
            _repo = repo;
            _limit = limit;
            _branch = branch;
            _referenceExpression = referenceExpression;
            _keyword = keyword;
        }

        internal static LogCmd Create(HgRepository repo) =>
            new(Preconditions.CheckNotNull(repo), 0, null, null, null);

        /// <summary>Limit the query to references that fit the <paramref name="referenceExpression"/>.</summary>
        internal LogCmd WithReferenceExpression(string referenceExpression)
        {
            if (string.IsNullOrEmpty(referenceExpression.Trim()))
            {
                throw new RepoException("Cannot log null or empty reference");
            }

            return new LogCmd(_repo, _limit, _branch, referenceExpression.Trim(), _keyword);
        }

        /// <summary>Limit the query to <paramref name="limit"/> results. Should be &gt; 0.</summary>
        public LogCmd WithLimit(int limit)
        {
            Preconditions.CheckArgument(limit > 0);
            return new LogCmd(_repo, limit, _branch, _referenceExpression, _keyword);
        }

        /// <summary>Only query for revisions from the branch <paramref name="branch"/>.</summary>
        internal LogCmd WithBranch(string branch) =>
            new(_repo, _limit, branch, _referenceExpression, _keyword);

        /// <summary>Only query for revisions with the keyword <paramref name="keyword"/>.</summary>
        internal LogCmd WithKeyword(string keyword) =>
            new(_repo, _limit, _branch, _referenceExpression, keyword);

        /// <summary>Run "hg log" and return zero or more <see cref="HgLogEntry"/>.</summary>
        public IReadOnlyList<HgLogEntry> Run()
        {
            var builder = ImmutableArray.CreateBuilder<string>();
            builder.Add("log");
            builder.Add("--verbose"); // verbose flag shows files in output

            if (_branch != null)
            {
                builder.Add("--branch");
                builder.Add(_branch);
            }

            // hg requires limit to be a positive integer
            if (_limit > 0)
            {
                builder.Add("--limit");
                builder.Add(_limit.ToString());
            }

            if (_referenceExpression != null)
            {
                builder.Add("--rev");
                builder.Add(_referenceExpression);
            }

            if (_keyword != null)
            {
                builder.Add("--keyword");
                builder.Add(_keyword);
            }

            builder.Add("-Tjson");
            try
            {
                CommandOutput output = _repo.Hg(_repo.GetHgDir(), builder.ToImmutable(), CommandRunner.DefaultTimeout);
                return ParseLog(output.GetStdout());
            }
            catch (RepoException e)
            {
                if (e.Message is not null && UnknownRevision.IsMatch(e.Message))
                {
                    throw new ValidationException("Unknown revision: " + e.Message);
                }

                if (e.Message is not null && InvalidRefExpression.IsMatch(e.Message))
                {
                    throw new RepoException("Syntax error in reference expression: " + e.Message);
                }

                throw;
            }
        }

        private static IReadOnlyList<HgLogEntry> ParseLog(string log)
        {
            if (log.Length == 0)
            {
                return ImmutableArray<HgLogEntry>.Empty;
            }

            try
            {
                List<HgLogEntry>? logEntries =
                    JsonSerializer.Deserialize<List<HgLogEntry>>(log.Trim());
                return logEntries is null
                    ? ImmutableArray<HgLogEntry>.Empty
                    : logEntries.ToImmutableArray();
            }
            catch (JsonException e)
            {
                throw new RepoException($"Cannot parse log output: {e.Message}");
            }
        }
    }

    /// <summary>An object that represents a commit as returned by 'hg log'.</summary>
    public sealed class HgLogEntry
    {
        [JsonPropertyName("node")]
        public string GlobalId { get; set; } = string.Empty;

        [JsonPropertyName("parents")]
        public List<string> Parents { get; set; } = new();

        [JsonPropertyName("user")]
        public string? User { get; set; }

        [JsonPropertyName("date")]
        public List<JsonElement> CommitDate { get; set; } = new();

        [JsonPropertyName("branch")]
        public string? Branch { get; set; }

        [JsonPropertyName("desc")]
        public string? Description { get; set; }

        [JsonPropertyName("files")]
        public List<string> Files { get; set; } = new();

        public IReadOnlyList<string> GetParents() => Parents;

        public string? GetUser() => User;

        internal string GetGlobalId() => GlobalId;

        internal DateTimeOffset GetZonedDate()
        {
            // hg -Tjson emits date as [epochSeconds, tzOffsetSeconds]; the offset sign is inverted
            // relative to a standard UTC offset (mirrors the Java implementation).
            long epochSeconds = (long)ToDouble(CommitDate[0]);
            int tzOffset = (int)ToDouble(CommitDate[1]);
            var date = DateTimeOffset.FromUnixTimeSeconds(epochSeconds);
            var offset = TimeSpan.FromSeconds(-1 * tzOffset);
            return date.ToOffset(offset);
        }

        private static double ToDouble(JsonElement element) =>
            element.ValueKind == JsonValueKind.String
                ? double.Parse(element.GetString()!)
                : element.GetDouble();

        public string? GetBranch() => Branch;

        public string? GetDescription() => Description;

        public IReadOnlyList<string> GetFiles() => Files;
    }
}
